src/interfaces/controllers/deregister-trailer/deregister-trailer.ts

```diff
@@ -0,0 +1,54 @@
+import { Request, Response, NextFunction } from 'express';
+import * as domain from '../../../domain';
+import { DataAccess } from '../../../utils/data-access';
+import { IRequestHandler } from '../i-request-handler';
+import { deregisterTrailerValidator } from './deregister-trailer-validator';
+
+export class DeregisterTrailer implements IRequestHandler {
+  private validate(payload?: domain.DeregisterTrailer): string {
+    const errors = deregisterTrailerValidator.validate(payload, { abortEarly: false }).error;
+    return errors?.message;
+  }
+
+  public async call(req: Request, res: Response, next: NextFunction): Promise<void> {
+    const deregisterTrailer = req.body as domain.DeregisterTrailer;
+    const { trn } = req.params;
+
+    const errors = this.validate(deregisterTrailer);
+    if (errors) {
+      next(new domain.HTTPError(400, errors));
+      return;
+    }
+
+    try {
+      const { reasonForDeregistration, deregisterDate } = deregisterTrailer;
+      let existingTrailerRegistration = await this.checkForExistingTrailerRegistration(trn);
+      if (existingTrailerRegistration) {
+        existingTrailerRegistration = {
+          ...existingTrailerRegistration,
+          reasonForDeregistration,
+          deregisterDate,
+        };
+        await DataAccess.getInstance().upsertTrailerRegistration(existingTrailerRegistration);
```

Comment:
We may want to be a dependency on this file too for testability and later let's say if we were to change the DB we use, etc..